### PR TITLE
Adding the possibility of minting tokens

### PR DESCRIPTION
**Pull request by Marchand-Nicolas**
Adding the possibility of minting tokens with verification of a signature given by the server

### CODE REVIEW

File: src/badge/badge.cairo [hunk 1]
```diff
@@ -216,6 +226,96 @@ func locked{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(id:
...
+    //   pedersen_ptr(HashBuiltin*)
+    //   range_check_ptr
+    // Explicit args:
+    //   new_signer_public_key(felt): The new public key
+    // Returns:
+    //   None
+    alloc_locals;
+    let (caller_address) = get_caller_address();
+    let (contract_owner) = owner();
+    assert caller_address = contract_owner;
```
Comment (Bal7hazar) [inline src/badge/badge.cairo:314]: `Ownable.assert_only_owner();` standsfor this usage.

**Review by Bal7hazar — CHANGES_REQUESTED** [review]
- Signature verification must be defined in the library file instead of contract file which used to handle all the logic steps where the contract file used to aggregate the high level features and expose them
- Since the contract is not upgradable, we need another contract to handle all this stuff (a minter contract). The minter contract must be upgradable to avoid any risk, it must handle a `claim` external function that implement all the logic done in this pull request. Finally when we are sure that everything is fine and risk free, we will transfer the ownership of the badge contract from admin to the minter (it could be interesting to have a transfer badge ownership function in the minter itself to send it back to the admin just in case).
- In addition, we will need to have unit tests to test the logic implemented in the library files and integration tests which virtually deploy the contracts and test interactions between them.

File: src/badge/Ibadge.cairo [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+from starkware.cairo.common.uint256 import Uint256
+
+@contract_interface
+namespace IBadgeContract {
+    func transferOwnership(newOwner: felt) {
+    }
+
+    func mint(to: felt, id: Uint256, amount: Uint256, data_len: felt, data: felt*) {
+    }
+}
```
Comment (Bal7hazar) [inline src/badge/Ibadge.cairo:13]: Interface files must be placed in src/interfaces folder

File: src/badge/minter.cairo [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+@storage_var
+func badge_contract_address() -> (badge_contract_address: felt) {
+}
+
+//
+// Constructor
+//
+
+@constructor
```
Comment (Bal7hazar) [inline src/badge/minter.cairo:38]: No constructor decorator for upgradable contracts, un peu de doc au besoin :
- https://docs.openzeppelin.com/contracts-cairo/0.5.0/proxies
- https://ctrlc03.github.io/

File: src/badge/minter.cairo [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    //   pedersen_ptr(HashBuiltin*)
+    //   range_check_ptr
+    // Explicit args:
+    //   new_implementation(felt): Address of the new implementation
+    // Returns:
+    //   None
+    Ownable.assert_only_owner();
+    Proxy._set_implementation_hash(new_implementation);
+    return ();
+}
```
Comment (Bal7hazar) [inline src/badge/minter.cairo:231]: I think you can add all other upgradable functions such as, setAdmin, getAdmin, getImplementationHash

**Review by Bal7hazar — CHANGES_REQUESTED** [review]

File: tests/units/badge/minter/test_initialization.cairo [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+// Starkware dependencies
+from starkware.cairo.common.cairo_builtins import HashBuiltin, SignatureBuiltin
+from starkware.cairo.common.alloc import alloc
+from starkware.cairo.common.math import assert_not_zero
+
+// Local dependencies
+from src.badge.minter import owner, initializer, getSignerPublicKey, getBadgeContractAddress
+
+@external
+func test_initialization{
```
Comment (cloudvenger) [inline tests/units/badge/minter/test_initialization.cairo:14]: You should do the same as it's done in `tests/units/minter/test_initialization.cairo`

File: tests/units/badge/minter/test_transferBadgeContractOwnership.cairo [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+from src.interfaces.badgeMinter import ICarbonableBadgeMinter
+from src.interfaces.badge import ICarbonableBadge
+
+@external
+func test_transferBadgeContractOwnership{
+    syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr, ecdsa_ptr : SignatureBuiltin*
+}() {
+    alloc_locals;
+
+    local minter_contract_address: felt;
```
Comment (cloudvenger) [inline tests/units/badge/minter/test_transferBadgeContractOwnership.cairo:21]: same as before for the `setup` function

**Review by cloudvenger — CHANGES_REQUESTED** [review]
It's missing all the integration tests.

You should take inspiration from the `tests/integrations/badge` directory and the `tests/integrations/protocol/test_minter.cairo` file

**Review by cloudvenger — CHANGES_REQUESTED** [review]
The part of code is clean ! It's cool @Marchand-Nicolas 

The architecture of tests is better, almost perfect.

I am sorry, their are still modification in tests (unit and integration) but the most difficult is done.

File: tests/integrations/badge/minter/library.cairo [hunk 7]
```diff
@@ -55,12 +55,27 @@ func setup{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
...
-        context.badge_minter.contract_address = deploy_contract("openzeppelin/upgrades/presets/Proxy.cairo", [context.badge_minter.class_hash, get_selector_from_name("initializer"), 4, context.admin_account_contract, context.badge_minter.public_key, context.badge.contract_address, 0]).contract_address
+        context.badge_minter.class_hash = declare(
+            contract=context.sources.badge_minter
+        ).class_hash
+        context.badge_minter.contract_address = deploy_contract(
+            contract=context.sources.proxy,
+            constructor_args={
+                "implementation_hash": context.badge_minter.class_hash,
+                "selector": context.selector.initializer,
+                "calldata": [context.admin_account_contract, context.badge_minter.public_key, context.badge.contract_address, 0]
```
Comment (cloudvenger) [inline tests/integrations/badge/minter/library.cairo:76]: as a comment but I will not have it modified for that.
You can also do like this:

```
"calldata": calldata.values(),
```
with above : 
```
 calldata = { ....}
```

**Review by cloudvenger — APPROVED** [review]

File: tests/units/badge/minter/test_claim.cairo [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        ids.badge_type = context.whitelist.badge_type
+    %}
+
+    // Prepare minter instance
+    prepare(minter_owner, public_key, carbonable_badge_contract_address, proxy_admin);
+
+    // Claim badge
+    %{ stop_prank_callable = start_prank(context.whitelist.whitelisted_user_address) %}
+    %{ mock_call(context.mocks.carbonable_badge_contract_address, "mint", [0]) %}
+    %{ mock_call(context.mocks.carbonable_badge_contract_address, "balanceOf", [0, 0]) %}
```
Comment (cloudvenger) [inline tests/units/badge/minter/test_claim.cairo:50]: Why, for a nominal case, the `mint` and `balanceOf` are to 0 ?

Comment (Marchand-Nicolas) [reply tests/units/badge/minter/test_claim.cairo:50]: For the mint it is a mistake on my part, it should just be []. For the balance, it is necessary to have not mint the NFT (and thus to have a balance of 0) for mint. I'll do an extra commit to fix the first point and add an explicative comment for the balance.

File: tests/integrations/badge/minter/library.cairo (hunk 7 above)
Comment (Marchand-Nicolas) [reply tests/integrations/badge/minter/library.cairo:76]: I will make the change, it will be cleaner

**Review by cloudvenger — APPROVED** [review]

**Review by cloudvenger — APPROVED** [review]

**Pull request merged**